### PR TITLE
add classification module to Detectron

### PR DESCRIPTION
**Pull request by bendidi**
add the possibility to use the conv backbone body and an MLP head to do image classification directly: 

- data input stays the same : coco format (added a script to convert from cifar100 to cls_coco format) 
- config file for classification in `configs/test_cls`
- Add X number of fully connected layers to the backbone conv

### CODE REVIEW

**Comment by EinavNamer** [comment]
Hello and thank you very much for your contribution.
How can I download "all changes" including all the other bug fixes you have done afterwards?
Thanks!

**Comment by bendidi** [comment]
@EinavNamer you an just directly clone the fork ? it's in here : https://github.com/Sterblue/Detectron.git

**Comment by EinavNamer** [comment]
OK thanks

**Comment by EinavNamer** [comment]
@bendidi - thanks again.
I'm having some trouble, hope you can help.
**With the following config file:**
MODEL:
  TYPE: generalized_rcnn
  CONV_BODY: ResNet.add_ResNet50_conv5_body
  NUM_CLASSES: 101
  CLASSIFICATION: True
NUM_GPUS: 1
SOLVER:
  WEIGHT_DECAY: 0.0001
  LR_POLICY: steps_with_decay
  BASE_LR: 0.01
  GAMMA: 0.1
  MAX_ITER: 20000
  STEPS: [0, 20000, 50000]
CLASSIFICATION:
  MLP_HEAD: classification_heads.add_Xmlp_head
  MLP_HEADS_DIM: [4096,4096]
  DROPOUT_RATE: 0.5
TRAIN:
  WEIGHTS: /home/einav/Projects/detectron/output/weighs/ImageNetPretrained/MSRA/R-50.pkl
  DATASETS: ('coco_2014_train',)
  SCALES: (32,)
  IMS_PER_BATCH: 8
  MAX_SIZE: 32
TEST:
  DATASETS: ('coco_2014_val',)
  SCALES: (32,)
  MAX_SIZE: 32
  PRECOMPUTED_PROPOSALS: False
OUTPUT_DIR: .

**I get this error:**
INFO detector.py: 436: Changing learning rate 0.000000 -> 0.003333 at iter 0
terminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at softmax_ops.cu:320] T.size() == N. 37 vs 8 Error from operator: 
input: "gpu_0/logits" input: "gpu_0/labels_int32" output: "gpu_0/cls_prob" output: "gpu_0/loss_cls" name: "" type: "SoftmaxWithLoss" arg { name: "scale" f: 1 } device_option { device_type: 1 cuda_gpu_id: 0 }
*** Aborted at 1523447369 (unix time) try "date -d @1523447369" if you are using GNU date ***
PC: @     0x7fcb4452a428 gsignal
*** SIGABRT (@0x3ea000014d9) received by PID 5337 (TID 0x7fc9f17fa700) from PID 5337; stack trace: ***
    @     0x7fcb448d0390 (unknown)
    @     0x7fcb4452a428 gsignal
    @     0x7fcb4452c02a abort
    @     0x7fcb35fb084d __gnu_cxx::__verbose_terminate_handler()
    @     0x7fcb35fae6b6 (unknown)
    @     0x7fcb35fae701 std::terminate()
    @     0x7fcb35fd9d38 (unknown)
    @     0x7fcb448c66ba start_thread
    @     0x7fcb445fc41d clone
    @                0x0 (unknown)

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
-----------------------------------------

**Here are my questions:**
(1) In the error, you can see that "T.size() == N. 37 vs 8" --> where 8 is the IMS_PER_BATCH: 8
Do you know how to fix the error?
(2) If we run dataset COCO-eval&test, why do you use NUM_CLASSES: 101 (cifar100)? why not 81?
(3) what is that parameter: MLP_HEADS_DIM: [4096,4096] ?
Based on the ResNet papaer(https://arxiv.org/pdf/1512.03385.pdf), when adding a classification head, the sizes should be 2048 (last conv5 dim ) --> 81 (num of COCO classes).

Thanks in advance,
Einav

**Comment by bendidi** [comment]
Hi , 

1. I just retested with cifar dataset and couldn't re the error (using a batch size of 258 ) : 
```
........
INFO net.py: 240: res5_2_branch2a             : (256, 512, 1, 1)     => res5_2_branch2a             : (256, 512, 1, 1)     ------- (op: AffineChannel)
INFO net.py: 240: res5_2_branch2a             : (256, 512, 1, 1)     => res5_2_branch2a             : (256, 512, 1, 1)     ------- (op: Relu)
INFO net.py: 240: res5_2_branch2a             : (256, 512, 1, 1)     => res5_2_branch2b             : (256, 512, 1, 1)     ------- (op: Conv)
INFO net.py: 240: res5_2_branch2b             : (256, 512, 1, 1)     => res5_2_branch2b             : (256, 512, 1, 1)     ------- (op: AffineChannel)
INFO net.py: 240: res5_2_branch2b             : (256, 512, 1, 1)     => res5_2_branch2b             : (256, 512, 1, 1)     ------- (op: Relu)
INFO net.py: 240: res5_2_branch2b             : (256, 512, 1, 1)     => res5_2_branch2c             : (256, 2048, 1, 1)    ------- (op: Conv)
INFO net.py: 240: res5_2_branch2c             : (256, 2048, 1, 1)    => res5_2_branch2c_bn          : (256, 2048, 1, 1)    ------- (op: AffineChannel)
INFO net.py: 240: res5_2_branch2c_bn          : (256, 2048, 1, 1)    => res5_2_sum                  : (256, 2048, 1, 1)    ------- (op: Sum)
INFO net.py: 240: res5_1_branch2c_bn          : (256, 2048, 1, 1)    => res5_2_sum                  : (256, 2048, 1, 1)    ------|
INFO net.py: 240: res5_2_sum                  : (256, 2048, 1, 1)    => res5_2_sum                  : (256, 2048, 1, 1)    ------- (op: Relu)
INFO net.py: 240: res5_2_sum                  : (256, 2048, 1, 1)    => final_avg                   : (256, 2048, 1, 1)    ------- (op: AveragePool)
INFO net.py: 240: final_avg                   : (256, 2048, 1, 1)    => fc6                         : (256, 4096)          ------- (op: FC)
INFO net.py: 240: fc6                         : (256, 4096)          => fc6                         : (256, 4096)          ------- (op: Relu)
INFO net.py: 240: fc6                         : (256, 4096)          => drop_fc6                    : (256, 4096)          ------- (op: Dropout)
INFO net.py: 240: drop_fc6                    : (256, 4096)          => fc7                         : (256, 4096)          ------- (op: FC)
INFO net.py: 240: fc7                         : (256, 4096)          => fc7                         : (256, 4096)          ------- (op: Relu)
INFO net.py: 240: fc7                         : (256, 4096)          => drop_fc7                    : (256, 4096)          ------- (op: Dropout)
INFO net.py: 240: drop_fc7                    : (256, 4096)          => logits                      : (256, 101)           ------- (op: FC)
INFO net.py: 240: logits                      : (256, 101)           => cls_prob                    : (256, 101)           ------- (op: SoftmaxWithLoss)
INFO net.py: 240: labels_int32                : (256,)               => cls_prob                    : (256, 101)           ------|
INFO net.py: 240: cls_prob                    : (256, 101)           => accuracy_cls                : ()                   ------- (op: Accuracy)
INFO net.py: 240: labels_int32                : (256,)               => accuracy_cls                : ()                   ------|
INFO net.py: 244: End of model: generalized_rcnn
json_stats: {"accuracy_cls": 0.007812, "eta": "11:04:38", "iter": 0, "loss": 5.543577, "loss_cls": 5.543577, "lr": 0.003333, "mb_qsize": 64, "mem": 1472, "time": 1.993948}
json_stats: {"accuracy_cls": 0.015625, "eta": "1:29:58", "iter": 20, "loss": 4.610148, "loss_cls": 4.610148, "lr": 0.003600, "mb_qsize": 43, "mem": 1472, "time": 0.270170}
json_stats: {"accuracy_cls": 0.018555, "eta": "0:58:13", "iter": 40, "loss": 4.517654, "loss_cls": 4.517654, "lr": 0.003867, "mb_qsize": 3, "mem": 1472, "time": 0.175003}
```
i'm testing with cifar100 dataset , to check if i could reproduce results from the papers, I added a script in tools to convert cifar100 to a pseudo coco format and use (mostly) the same  input pipeline of Detectron

2. i'm using cifar100 dataset, so it's 101 classes but you can change it as you wish for any custom dataset 

3. `MLP_HEADS_DIM` is the number of fully connected layers and their size to add after the last conv layer, you can just use an empty array to not add any FC layer and have  `2048 -> NUM_CLASSES`

hope this helps , cheers

**Comment by rbgirshick** [comment]
Thank you for the PR, however we don't plan to add whole-image classification support to detectron.
